In [1]:
import pandas as pd
import numpy as np

# Question 1

In [2]:
raw = pd.read_csv('2019 Winter Data Science Intern Challenge Data Set - Sheet1.csv')

In [18]:
raw['average_item_amount'] = raw['order_amount']/raw['total_items']

In [19]:
raw.head()

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

            created_at  average_item_amount  
0  2017-03-13 12:36:56                112.0  
1  2017-03-03 17:38:52                 90.0  
2   2017-03-14 4:23:56                144.0  
3  2017-03-26 12:43:37                156.0  
4   2017-03-01 4:35:11                156.0

In [20]:
raw.shape

(5000, 8)

In [24]:
shops = raw.groupby(['shop_id', 'average_item_amount']).sum()[['order_amount', 'total_items']].reset_index()

In [25]:
shops.head()

shop_id  average_item_amount  order_amount  total_items
0        1                158.0         13588           86
1        2                 94.0          9588          102
2        3                148.0         14652           99
3        4                128.0         13184          103
4        5                142.0         13064           92

In [26]:
shops.shape

(100, 4)

The above confirms that the data contains 100 shops and each of these sell 1 product (assuming different products have different prices)

In [32]:
round(raw['order_amount'].mean(), 2)

3145.13

This confirms there were no technical issues in the calculation of the mean

In [48]:
shops.sort_values(by='order_amount', ascending=False).head(10)

shop_id  average_item_amount  order_amount  total_items
41       42                352.0      11990176        34063
77       78              25725.0       2263800           88
88       89                196.0         23128          118
80       81                177.0         22656          128
5         6                187.0         22627          121
12       13                160.0         21760          136
58       59                178.0         21538          121
70       71                164.0         21320          130
18       19                163.0         20538          126
69       70                173.0         20241          117

In [49]:
shops.sort_values(by='total_items', ascending=False).head(10)

shop_id  average_item_amount  order_amount  total_items
41       42                352.0      11990176        34063
12       13                160.0         21760          136
83       84                153.0         20196          132
70       71                164.0         21320          130
52       53                112.0         14560          130
80       81                177.0         22656          128
40       41                118.0         14986          127
23       24                140.0         17640          126
18       19                163.0         20538          126
13       14                116.0         14036          121

In [50]:
shops.sort_values(by='average_item_amount', ascending=False).head(10)

shop_id  average_item_amount  order_amount  total_items
77       78              25725.0       2263800           88
41       42                352.0      11990176        34063
11       12                201.0         18693           93
88       89                196.0         23128          118
98       99                195.0         18330           94
49       50                193.0         17756           92
37       38                190.0         13680           72
50       51                187.0         16643           89
5         6                187.0         22627          121
10       11                184.0         17480           95

### a. 

Looking at these three tables, shops 42 and 78 clearly stand out from the rest. Shop 42 has a total_items value far larger than any other shop and the price of shop 78's sneaker is much higher than any other store's sneaker. So while the average amount of 3145.13 calculated earlier is not technically wrong, that number is being skewed by these two stores. The question now is whether it makes sense to keep the data for these two stores in this analysis. 

In [56]:
raw[raw['user_id']==607]

order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
520        521       42      607        704000         2000    credit_card   
1104      1105       42      607        704000         2000    credit_card   
1362      1363       42      607        704000         2000    credit_card   
1436      1437       42      607        704000         2000    credit_card   
1562      1563       42      607        704000         2000    credit_card   
1602      1603       42      607        704000         2000    credit_card   
2153      2154       42      607        704000         2000    credit_card   
2297      2298       42      607        704000         2000    credit_card   
2835      2836       42      607        704000         2000    credit_card   
2969      2970       42      607        704000         2000    credit_card   
3332      3333       42      607        704000         2000    credit_card   
4056      4057       42      607        704000         2000    credit_card   
4646      4647       42      607        704000         2000    credit_card   
4868      4869       42      607        704000         2000    credit_card   
4882      4883       42      607        704000         2000    credit_card   

              created_at  average_item_amount  
15    2017-03-07 4:00:00                352.0  
60    2017-03-04 4:00:00                352.0  
520   2017-03-02 4:00:00                352.0  
1104  2017-03-24 4:00:00                352.0  
1362  2017-03-15 4:00:00                352.0  
1436  2017-03-11 4:00:00                352.0  
1562  2017-03-19 4:00:00                352.0  
1602  2017-03-17 4:00:00                352.0  
2153  2017-03-12 4:00:00                352.0  
2297  2017-03-07 4:00:00                352.0  
2835  2017-03-28 4:00:00                352.0  
2969  2017-03-28 4:00:00                352.0  
3332  2017-03-24 4:00:00                352.0  
4056  2017-03-28 4:00:00                352.0  
4646  2017-03-02 4:00:00                352.0  
4868  2017-03-22 4:00:00                352.0  
4882  2017-03-25 4:00:00                352.0

In [55]:
raw[(raw['shop_id']==42) & (raw['user_id']!=607)]

order_id  shop_id  user_id  order_amount  total_items payment_method  \
40          41       42      793           352            1    credit_card   
308        309       42      770           352            1    credit_card   
409        410       42      904           704            2    credit_card   
834        835       42      792           352            1           cash   
835        836       42      819           704            2           cash   
938        939       42      808          1056            3    credit_card   
979        980       42      744           352            1          debit   
1364      1365       42      797          1760            5           cash   
1367      1368       42      926          1408            4           cash   
1471      1472       42      907          1408            4          debit   
1512      1513       42      946           352            1          debit   
1520      1521       42      756           704            2          debit   
1911      1912       42      739           704            2           cash   
1929      1930       42      770           352            1    credit_card   
2003      2004       42      934           704            2           cash   
2018      2019       42      739           352            1          debit   
2053      2054       42      951           352            1          debit   
2273      2274       42      747           704            2          debit   
2491      2492       42      868           704            2          debit   
2609      2610       42      868           704            2          debit   
2766      2767       42      970           704            2    credit_card   
2987      2988       42      819          1056            3           cash   
3513      3514       42      726          1056            3          debit   
3651      3652       42      830           352            1    credit_card   
3697      3698       42      839           352            1          debit   
3903      3904       42      975           352            1          debit   
3998      3999       42      886           352            1          debit   
4231      4232       42      962           352            1           cash   
4294      4295       42      859           704            2           cash   
4326      4327       42      788           704            2          debit   
4421      4422       42      736           704            2    credit_card   
4625      4626       42      809           352            1    credit_card   
4745      4746       42      872           352            1          debit   
4767      4768       42      720           704            2    credit_card   

               created_at  average_item_amount  
40    2017-03-24 14:15:41                352.0  
308   2017-03-11 18:14:39                352.0  
409   2017-03-04 14:32:58                352.0  
834   2017-03-25 21:31:25                352.0  
835   2017-03-09 14:15:15                352.0  
938   2017-03-13 23:43:45                352.0  
979   2017-03-12 13:09:04                352.0  
1364   2017-03-10 6:28:21                352.0  
1367   2017-03-13 2:38:34                352.0  
1471  2017-03-12 23:00:22                352.0  
1512  2017-03-24 13:35:04                352.0  
1520  2017-03-22 13:10:31                352.0  
1911   2017-03-07 5:42:52                352.0  
1929   2017-03-17 8:11:13                352.0  
2003   2017-03-26 9:21:26                352.0  
2018  2017-03-01 12:42:26                352.0  
2053  2017-03-19 11:49:12                352.0  
2273  2017-03-27 20:48:19                352.0  
2491  2017-03-01 18:33:33                352.0  
2609  2017-03-23 18:10:14                352.0  
2766  2017-03-05 10:45:42                352.0  
2987   2017-03-03 9:09:25                352.0  
3513  2017-03-24 17:51:05                352.0  
3651  2017-03-24 22:26:58                352.0  
3697   2017-03-12 2:45:09                352.0  
3

In the case of shop 42, it seems there is one consumer with user_id 607 buying shop 42's sneakers 2000 items at a time. No rational consumer would buy sneakers at this rate so it is fair to assume this buyer is either another business reselling this sneaker or this data is the result of a bug in the pipeline. If the first case is true, then this data belongs in a separate b2b sales dataset and if the second case is true then this data does not belong in any dataset. Hence from here on out I will ignore orders from user_id 607 at shop 42. 

In [57]:
raw = raw[raw['user_id']!=607]

In [58]:
shops = raw.groupby(['shop_id', 'average_item_amount']).sum()[['order_amount', 'total_items']].reset_index()

In [61]:
round(raw['order_amount'].mean(), 2)

754.09

Even withoutbefore doing anything to account for the shop 78 sneaker, removing purchases from user 607 reduced the mean order amount significantly to a more believable $754 

The situation with shop 78 is slightly different. Here the amount of sneakers being bought by customers is reasonable, no one customer is buying disproportionately more than the others. The issue here is that the price of their sneaker is far greater than any other store in the data. In this case however I do not feel comfortable removing these sales because $25725 for a pair of shoes is conceivable and not uncommon among luxury brands.

### b. & c.

In [62]:
round(raw['order_amount'].median(), 2)

284.0

Finally, calculating the median rather than the mean allows us to report a number that is a good representation of the average consumer without having to remove shop 78's sales from the data. And for the month of March 2017 this median was at $284

# Question 2

### a.

SELECT Shippers.ShipperName, \n
       COUNT(Shippers.ShipperName) AS OrdersShipped \
FROM Orders \
LEFT JOIN Shippers ON Orders.ShipperID=Shippers.ShipperID \
WHERE ShipperName='Speedy Express' \
GROUP BY Shippers.ShipperName;

Speedy Express Shipped 54 Orders

### b.

SELECT TOP 1 COUNT(Employees.EmployeeID) AS EmployeeOrders,\
Employees.EmployeeID,\
Employees.LastName\
FROM Orders\
LEFT JOIN Employees ON Orders.EmployeeID=Employees.EmployeeID\
GROUP BY Employees.EmployeeID, Employees.LastName\
ORDER BY COUNT(Employees.EmployeeID) DESC;

Peacock is the last name of the employee with the most orders

### c.

SELECT TOP 1 \
p.ProductID,\
p.ProductName,\
SUM(od.Quantity) AS TotalOrders \
FROM (((OrderDetails od\
LEFT JOIN Orders o ON  od.OrderID=o.OrderID)\
LEFT JOIN Customers c ON o.CustomerID=c.CustomerID)\
LEFT JOIN Products p ON od.ProductID=p.ProductID)\
WHERE Country='Germany'\
GROUP BY p.ProductID, p.ProductName \
ORDER BY SUM(od.Quantity) DESC;

Boston Crab Meat was ordered the most by customers in Germany